chore: remove usage of bitcoin_hashes::hex::ToHex trait

tvolk131 opened the pull request:
Replace usage of `bitcoin_hashes::hex::ToHex` trait with `hex::ToHex` since the former was [removed](https://github.com/rust-bitcoin/rust-bitcoin/pull/1531) in `bitcoin 0.30.0`. Also replace `bitcoin_hashes::hex::FromHex` with `hex::FromHex` where it is easy to do so, even though `bitcoin_hashes::hex::FromHex` still exists.

This is a small step towards updating to `bitcoin 0.30` (#2356).

justinmoon commented:
Rebasing to include https://github.com/fedimint/fedimint/pull/4005 once it's merged should fix flake CI job

m1sterc001guy commented on fedimint-core/src/admin_client.rs:
```diff
@@ -269,7 +269,7 @@ mod serde_tls_cert {
         let value: Cow<str> = Deserialize::deserialize(deserializer)?;
         Ok(rustls::Certificate(
-            Vec::from_hex(&value).map_err(D::Error::custom)?,
+            Vec::from_hex(&*value).map_err(D::Error::custom)?,
```
Can you just do `value` here? De-referencing then borrowing is weird

tvolk131 replied:
`value` doesn't work. However, I was able to replace the deref-borrow with `value.as_ref()`.

m1sterc001guy commented on fedimint-core/src/config.rs:
```diff
@@ -931,7 +931,7 @@ pub mod serde_binary_human_readable {
         if d.is_human_readable() {
             let hex_str: Cow<str> = Deserialize::deserialize(d)?;
-            let bytes = Vec::from_hex(&hex_str).map_err(serde::de::Error::custom)?;
+            let bytes = Vec::from_hex(&*hex_str).map_err(serde::de::Error::custom)?;
```
Same comment as above about de-referencing then borrowing

tvolk131 replied:
Same response as above (replaced with `value.as_ref()`)

m1sterc001guy commented on fedimint-core/src/txoproof.rs:
```diff
@@ -98,7 +98,7 @@ impl<'de> Deserialize<'de> for TxOutProof {
         if deserializer.is_human_readable() {
             let hex_str: Cow<str> = Deserialize::deserialize(deserializer)?;
-            let bytes = Vec::from_hex(&hex_str).map_err(D::Error::custom)?;
+            let bytes = Vec::from_hex(&*hex_str).map_err(D::Error::custom)?;
```
Same comment here

tvolk131 replied:
Same response as above (replaced with `value.as_ref()`)

m1sterc001guy commented on fedimint-server/src/config/mod.rs:
```diff
@@ -768,7 +768,7 @@ mod serde_tls_cert_map {
 
         for (key, value) in map {
-            let cert = rustls::Certificate(Vec::from_hex(&value).map_err(D::Error::custom)?);
+            let cert = rustls::Certificate(Vec::from_hex(&*value).map_err(D::Error::custom)?);
```
Same comment here

tvolk131 replied:
Same response as above (replaced with `value.as_ref()`)

m1sterc001guy commented on fedimint-server/src/config/mod.rs:
```diff
@@ -795,7 +795,7 @@ mod serde_tls_key {
     {
         let hex_str: Cow<str> = Deserialize::deserialize(deserializer)?;
-        let bytes = Vec::from_hex(&hex_str).map_err(serde::de::Error::custom)?;
+        let bytes = Vec::from_hex(&*hex_str).map_err(serde::de::Error::custom)?;
```
Same comment here

tvolk131 replied:
Same response as above (replaced with `value.as_ref()`)

m1sterc001guy commented on gateway/ln-gateway/src/rpc/mod.rs:
```diff
@@ -197,7 +197,7 @@ pub fn serde_hex_deserialize<'d, T: bitcoin::consensus::Decodable, D: Deserializ
     if d.is_human_readable() {
         let hex_str: Cow<str> = Deserialize::deserialize(d)?;
-        let bytes = Vec::from_hex(&hex_str).map_err(serde::de::Error::custom)?;
+        let bytes = Vec::from_hex(&*hex_str).map_err(serde::de::Error::custom)?;
```
Same comment here

tvolk131 replied:
Same response as above (replaced with `value.as_ref()`)

m1sterc001guy commented:
ACK, I did a similar change in [my branch](https://github.com/m1sterc001guy/fedimint/tree/update_rust_bitcoin), I think its a good step. Added some nit comments.

I believe when have a git hook or something that prevents us from using certain libraries and rust code. Should we add `to_hex` from `bitcoin_hashes` to this hook?

bradleystachurski reviewed (commented):
CI is failing to observe the correct txid in the mempool when submitting a withdrawal. I've tested locally and replicated. Adding additional logs for the transactions in the mempool shows the txid returned from `fedimint-cli withdraw` ([src](https://github.com/fedimint/fedimint/blob/4c2ca345dab8a6614c836b9aa0aaefd3e3497330/devimint/src/main.rs#L802)) doesn't match what's in the mempool.

```
2024-01-08T19:44:37.276315Z  INFO devimint: txid=e23882174285e8fff9d923851801aa5689a45d72dfb78bad68e9550680c60504
2024-01-08T19:44:38.279650Z  INFO devimint: mempool_res=Ok([0405c6800655e968ad8bb7df725da48956aa01188523d9f9ffe88542178238e2])
```

I've pushed a debugging branch if you want to replicate on your end using `./scripts/tests/devimint-cli-test.sh`.
https://github.com/tvolk131/fedimint/compare/remove_bitcoin_hashes_hex...bradleystachurski:fedimint:debug-hash

elsirion commented:
Some hashes need to be parsed/serialized in reverse order in Bitcoin, `bitcoin-hashes` should do it correctly, if you use something else it might need manual reversal.

dpc commented:
@tvolk131 ping. Do you still have time to wrap it up? Would be a shame if it didn't get in. I think Eric is right and this is about reversed encoding (you can even see it in the ones posted by @bradleystachurski).

tvolk131 commented:
> Do you still have time to wrap it up?

Yes! I'll get around to this in the next few days. Would love to get this merged!

justinmoon commented:
Re-running CI. Perhaps a rebase would help with CI flakes.

elsirion commented:
Needs rebase :cry:

douglaz commented:
Another rebase? :see_no_evil:

elsirion reviewed (approved):
Tests run through and code looks sensible, so appears to be alright. How did you solve the problem with serialization order of bytes for certain Bitcoin hashes?

tvolk131 commented:
> How did you solve the problem with serialization order of bytes for certain Bitcoin hashes?

The issue was in `fedimint-cli/src/client.rs`. In the broken code I had `"txid": txid.encode_hex::<String>()` but I fixed it by changing it to `"txid": format!("{txid}"),`. Apologies for the confusion! I rebased and squashed which was a bad idea in retrospect.

douglaz commented on fedimint-cli/src/client.rs:
```diff
@@ -581,7 +580,7 @@ pub async fn handle_command(
                     WithdrawState::Succeeded(txid) => {
                         return Ok(json!({
-                            "txid": txid.to_hex(),
+                            "txid": format!("{txid}"),
```
so it's possible to incorrectly write `txid.encode_hex::<String>()`? If so we should try to do something to avoid such traps

elsirion replied:
I implemented our `consensus_encode_to_hex` correctly and replaced the `format!` with it. There not much we can do about a foreign function incorrectly encoding a foreign type unfortunately.

tvolk131 commented:
Needed another rebase 😞 hoping to get this merged asap!

justinmoon reviewed (approved):
re-approving on behalf of @elsirion and @m1sterc001guy so we can get this merged

tvolk131 commented:
😬🤞
